compiler/rustc_mir/src/transform/flatten_locals.rs

```diff
@@ -0,0 +1,211 @@
+use crate::transform::MirPass;
+use rustc_data_structures::fx::{FxHashMap, FxHashSet};
+use rustc_index::vec::{Idx, IndexVec};
+use rustc_middle::mir::visit::*;
+use rustc_middle::mir::*;
+use rustc_middle::ty::TyCtxt;
+use std::collections::hash_map::Entry;
+
+pub struct FlattenLocals;
+
+impl<'tcx> MirPass<'tcx> for FlattenLocals {
+    fn run_pass(&self, tcx: TyCtxt<'tcx>, body: &mut Body<'tcx>) {
+        if tcx.sess.mir_opt_level() < 4 {
+            return;
+        }
+
+        let replacements = compute_flattening(tcx, body);
+        let mut all_dead_locals = FxHashSet::default();
+        all_dead_locals.extend(replacements.fields.keys().map(|p| p.local));
+        if all_dead_locals.is_empty() {
+            return;
+        }
+
+        ReplacementVisitor { tcx, map: &replacements, all_dead_locals }.visit_body(body);
+
+        let mut replaced_locals: IndexVec<_, _> = IndexVec::new();
+        for (k, v) in replacements.fields {
+            replaced_locals.ensure_contains_elem(k.local, || Vec::new());
+            replaced_locals[k.local].push(v)
+        }
+        // Sort locals to avoid depending on FxHashMap order.
+        for v in replaced_locals.iter_mut() {
+            v.sort_unstable()
+        }
+        for bbdata in body.basic_blocks_mut().iter_mut() {
+            bbdata.expand_statements(|stmt| {
+                let source_info = stmt.source_info;
+                let (live, origin_local) = match &stmt.kind {
+                    StatementKind::StorageLive(l) => (true, *l),
+                    StatementKind::StorageDead(l) => (false, *l),
+                    _ => return None,
+                };
+                replaced_locals.get(origin_local).map(move |final_locals| {
```

Comment:
This removes storage markers for locals without any replacements (`replaced_locals` is an `IndexVec` now, so `get` doesn't have desired effect).